staking/contracts/RariMineV3.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: MIT
+
+/**
+ *Submitted for verification at Etherscan.io on 2020-07-20
+*/
+
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+import "./libs/LibString.sol";
+import "./libs/LibAddress.sol";
+import "./libs/LibUint.sol";
+import "./libs/LibStakingMath.sol";
+import "./IRariMine.sol";
+import "./IStaking.sol";
+import "./IERC20Read.sol";
+
+contract RariMineV3 is OwnableUpgradeable, IRariMine, IERC20Read {
+    using SafeMathUpgradeable for uint;
+    using LibString for string;
+    using LibUint for uint;
+    using LibAddress for address;
+
+    IERC20Upgradeable public token;
+    address           public tokenOwner;
+    IStaking          public staking;
+
+    string   private _name;
+    string   private _symbol;
+    uint8    private _decimals;
+    uint256  private _totalSupply;
+
+    uint256 constant CLAIM_FORMULA_STAKE   = 60000000;  // 60% to stake
+    uint256 constant CLAIM_FORMULA_CLAIM   = 40000000;  // 40% to withdraw
+    uint256 constant CLAIM_FORMULA_DIVIDER = 100000000; //  
+    uint256 constant CLAIM_CLIFF_WEEKS     = 42;        // the meaning of life, the universe, and everything
+    uint256 constant CLAIM_SLOPE_WEEKS     = 42;        // the meaning of life, the universe, and everything
+    uint8   public constant VERSION               = 1;
+
+    mapping(address => uint) public claimed;
+
+    function __RariMineV3_init(IERC20Upgradeable _token, address _tokenOwner, IStaking _staking) external initializer {
+        __RariMineV3_init_unchained(_token, _tokenOwner, _staking);
+        __Ownable_init_unchained();
+        __Context_init_unchained();
+    }
+
+    function __RariMineV3_init_unchained(IERC20Upgradeable _token, address _tokenOwner, IStaking _staking) internal initializer {
+        token = _token;
+        tokenOwner = _tokenOwner;
+        staking = _staking;
+        _decimals = 18;
+        _name = "Rari Mine Claimed Tokens";
+        _symbol = "RariMineCT";
+    }
+
+    function claim(Balance memory _balance, uint8 v, bytes32 r, bytes32 s) public {
+        require(prepareMessage(_balance, address(this)).recover(v, r, s) == owner(), "owner should sign balances");
+
+        address recipient = _balance.recipient;
+        if (_msgSender() == recipient) {
+            uint toClaim = _balance.value.sub(claimed[recipient]);
+            require(toClaim > 0, "nothing to claim");
+            claimed[recipient] += _balance.value;
+            _totalSupply += _balance.value;
+
+            // claim rari tokens
+            uint claimAmount = toClaim.mul(CLAIM_FORMULA_CLAIM).div(CLAIM_FORMULA_DIVIDER);
```

Comment:
We need only one param, also you can set in basis points, 10000 divider

Reply:
have a set of parameters which owner can control, removed unused variables